File: Data Structures & Algorithms/non-cyclical-number/submission-3.cpp
class Solution {
public:
    bool isHappy(int n) {
        unordered_set<int> s;
        while (n != 1) {
            if (s.contains(n)) {
                return false;
            }
            s.insert(n);
            n = replace(n);
        }
        return true;
    }

    int replace(int n) {
        int ret = 0;
        while (n) {
            int val = n % 10;
            n /= 10;
            ret += pow(val, 2);
        }
        return ret;
    }
};

File: Data Structures & Algorithms/decode-ways/submission-3.py
class Solution:
    def numDecodings(self, s: str) -> int:
        dp = {}

        def dfs(i):
            ret = 0
            if i in dp:
                return dp[i]

            if i == len(s):
                return 1

            if s[i] == '0':
                return 0
            
            if i < len(s):
                ret += dfs(i + 1)
            
            if i < len(s) - 1 and 1 <= int(s[i: i + 2]) <= 26:
                ret += dfs(i + 2)
            
            dp[i] = ret
            return ret

        return dfs(0)